src/main/java/com/aws/greengrass/shadowmanager/sync/model/MergedFullShadowSyncRequest.java

```diff
@@ -0,0 +1,143 @@
+            // If a full sync has been directly requested, not due to merging of requests, respect it.
+            super.execute(context);
+            return;
+        }
+
+        List<SyncRequest> necessaryMergedUpdates = getNecessaryMergedRequests(context);
+        if (necessaryMergedUpdates.isEmpty()) {
+            logger.atDebug()
+                    .kv(LOG_THING_NAME_KEY, getThingName())
+                    .kv(LOG_SHADOW_NAME_KEY, getShadowName())
+                    .log("Ignoring unnecessary sync request");
+            return;
+        }
+
+        if (necessaryMergedUpdates.stream().allMatch(r -> r instanceof CloudUpdateSyncRequest)
+                || necessaryMergedUpdates.stream().allMatch(r -> r instanceof LocalUpdateSyncRequest)) {
+            SyncRequest consolidatedUpdateRequest = necessaryMergedUpdates.stream().reduce(merger::merge).get();
+            logger.atDebug()
+                    .kv(LOG_THING_NAME_KEY, getThingName())
+                    .kv(LOG_SHADOW_NAME_KEY, getShadowName())
+                    .log("Full sync not needed, executing a "
+                            + consolidatedUpdateRequest.getClass().getSimpleName() + " instead");
+            consolidatedUpdateRequest.execute(context);
+            return;
+        }
+
+        super.execute(context);
+    }
+
+    /**
+     * Create a list of all the merged requests that require execution,
+     * as deemed by {@link BaseSyncRequest#isUpdateNecessary(SyncContext)}.
+     *
+     * @param context sync context
+     * @return sync requests
+     * @throws RetryableException       When error occurs in sync operation indicating a request needs to be retried
+     * @throws SkipSyncRequestException When error occurs in sync operation indicating a request needs to be skipped.
+     * @throws UnknownShadowException   When shadow not found in the sync table.
+     */
+    private List<SyncRequest> getNecessaryMergedRequests(SyncContext context)
+            throws RetryableException, UnknownShadowException, SkipSyncRequestException {
+        List<SyncRequest> necessaryUpdates = new ArrayList<>();
+        for (SyncRequest request : mergedRequests) {
+            if (!(request instanceof BaseSyncRequest)) {
+                continue;
+            }
+            if (((BaseSyncRequest) request).isUpdateNecessary(context)) {
+                necessaryUpdates.add(request);
+            }
+        }
+        return necessaryUpdates;
+    }
+
+    /**
+     * Returns a new list that matches the input, but all {@link MergedFullShadowSyncRequest}s
+     * are replaced (flat-mapped) with their {@link MergedFullShadowSyncRequest#mergedRequests}.
+     *
+     * @param mergedRequests sync requests
+     * @return sync requests
+     */
+    private static List<SyncRequest> flatten(List<SyncRequest> mergedRequests) {
+        return mergedRequests.stream()
+                .flatMap(r -> {
+                    if (r instanceof MergedFullShadowSyncRequest) {
```

Comment:
nit: I generally try to avoid nesting. We could
```suggestion
    if (!r instanceof MergedFullShadowSyncRequest) {
         return Stream.of(r);
    }
    
    MergedFullShadowSyncRequest other = (MergedFullShadowSyncRequest) r;
      if (other.getMergedRequests() == null) {
          return Stream.empty();
      } else {
          return other.getMergedRequests().stream();
      }
```

Reply:
👍  will do